Docs: Time series dimensions and labels (#26299)
grafana/grafana: docs/sources/getting-started/timeseries-dimensions.md

+++
title = "Time series dimensions"
description = "time series dimensions"
keywords = ["grafana", "intro", "guide", "concepts", "timeseries", "labels"]
type = "docs"
aliases = ["/docs/grafana/latest/guides/timeseries-dimensions"]
[menu.docs]
name = "Time series dimensions"
identifier = "time_series_dimensions"
parent = "guides"
weight = 300
+++

# Time series dimensions

> In at the end of the ["Time series databases" section of "Introduction to time series"]({{< relref "timeseries.md#time-series-databases" >}}) the concept of Labels (a.k.a tags) is introduced:

Reviewer: ```suggestion
At the end of the ["Time series databases" section of "Introduction to time series"]({{< relref "timeseries.md#time-series-databases" >}}), the concept of _labels_, also called  _tags_, is introduced:
```

> Another feature of a TSDB is the ability to filter measurements using _tags_. Each data point is labeled with a tag that adds context information, such as where the measurement was taken. ...

With time series data, the data often contains more than a single series, and is a set of multiple time series. Many Grafana data sources support this type of data.

{{< docs-imagebox img="/img/docs/example_graph_multi_dim.png" class="docs-image--no-shadow" max-width="850px" >}}

> The common case is issuing a single query for a measurement with one or more additional properties as dimensions. For example querying a temperature measurement along with a location property as a dimension of the measurement. In this case, multiple series are returned back from that single query. Each series in this case has unique location. To uniquely identify these series within a set of time series, Grafana stores this information in _Labels_.

Reviewer: ```suggestion
The common case is issuing a single query for a measurement with one or more additional properties as dimensions. For example, querying a temperature measurement along with a location property as a dimension of the measurement. In this case, multiple series are returned back from that single query. Each series in this case has unique location. To uniquely identify these series within a set of time series, Grafana stores this information in _labels_.
```

Reviewer: When writing for the web, avoid making sentences that are more than 25 words long or paragraphs more than three sentences long. Please break this paragraph into smaller ones.

For the example, consider adding a markdown table to illustrate your point.

## Labels

> Each time series in Grafana optionally has Labels. Labels are set a of key/value pairs for identifying dimensions. Example Labels could are `{location=us}` or `{country=us,state=ma,city=boston}`. Within a set of time series, the combination of its name and labels identifies each series. For example, `temperature {country=us,state=ma,city=boston}`.

Reviewer: Unless referring to something in the UI that is capitalized, lowercase "label" throughout.

> Different sources of time series data have dimensions stored natively, or common storage patterns that allow the data to be extracted into dimensions. Time series databases (TSDBs) usually natively support dimensionality. In TSDBs such as Graphite or OpenTSDB the term _Tags_ is used - in Prometheus the same term as used as in Grafana - _Labels_. In table databases such SQL, these dimensions are generally the `GROUP BY` parameters of a query.

Reviewer: Lowercase "tags" and "labels." See earlier comment about paragraph length.

Also, you are using hyphens instead of em dashes. Please replace them with em dashes or rewrite the sentence so they are not needed.

> By storing dimensionality as Labels for time series in Grafana it allows for the possibility of data transformations and alerting to also support multiple dimensions.

Reviewer: ```suggestion
By storing dimensionality as labels for time series in Grafana, you can transform data and alert across multiple dimensions.
```

Author: Oh I committed this but it isn't true. This is something we will be able to do in the future, but Grafana isn't there yet. For alerting it will be nextgen alerting. Not sure how to say that - or maybe this should be left out alltogether?

Reviewer: I'd say leave it out. Otherwise we'll get people asking when this will be applied. Forever.

## Multiple dimensions in table format

> In SQL or SQL-like databases that return table responses, additional dimensions are generally found as columns in the table that is returned as a response from the query.

Reviewer: ```suggestion
In SQL or SQL-like databases that return table responses, additional dimensions usually columns in the query response table.
```

### Single dimension

> For example, consider a pseudo query like:

Reviewer: I have never heard of a "pseudo query." Will this be understandable to your general audience?

Author: I meant Pseudocode SQL https://en.wikipedia.org/wiki/Pseudocode . But the point is just that it isn't any *actual* particular language  - so sort of "pretend" SQL.

But I think that point can be removed altogether -- I can just remove that word and put "consider a query like" given the scope/context of this document.

Reviewer: Yes, I agree.

```sql
SELECT BUCKET(StartTime, 1h), AVG(Temperature) AS Temp, Location FROM T
  GROUP BY BUCKET(StartTime, 1h), Location
  ORDER BY time asc
```

> Might return a table like with a 3 columns that each respectively have a type of Time, Number, and String:

Reviewer: ```suggestion
Might return a table like with three columns that each respectively have a type of Time, Number, and String:
```

| StartTime  | Temp | Location |
| ---------- | ---- | -------- |
| 09:00      | 24   | LGA      |
| 09:00      | 20   | BOS      |
| 10:00      | 26   | LGA      |
| 10:00      | 22   | BOS      |

> The table format is _Long_ formatted time series (aka Tall). It has repeated time stamps, and repeated values in Location. In this case, we have two time series in the set that would be identified as `Temp {Location=LGA}` and `Temp {Location=BOS}`.

Reviewer: ```suggestion
The table format is _long_ formatted time series, also called _tall_. It has repeated time stamps, and repeated values in Location. In this case, we have two time series in the set that would be identified as `Temp {Location=LGA}` and `Temp {Location=BOS}`.
```

> Individual time series from the set are extracted by using the time typed column `StartTime` as the time index of the time series, the numeric typed column `Temp` as the series Name, and the name and values of the string typed `Location` column to build the labels, such as Location=LGA.

Reviewer: ```suggestion
Individual time series from the set are extracted by using the time typed column `StartTime` as the time index of the time series, the numeric typed column `Temp` as the series name, and the name and values of the string typed `Location` column to build the labels, such as Location=LGA.
```

+++
title = "Time series dimensions"
description = "time series dimensions"
keywords = ["grafana", "intro", "guide", "concepts", "timeseries", "labels"]
type = "docs"
aliases = ["/docs/grafana/latest/guides/timeseries-dimensions"]
[menu.docs]
name = "Time series dimensions"
identifier = "time_series_dimensions"
parent = "guides"
weight = 300
+++

# Time series dimensions

At the end of the ["Time series databases" section of "Introduction to time series"]({{< relref "timeseries.md#time-series-databases" >}}), the concept of _labels_, also called  _tags_, is introduced:

> Another feature of a TSDB is the ability to filter measurements using _tags_. Each data point is labeled with a tag that adds context information, such as where the measurement was taken. ...

> With time series data, the data often contains more than a single series, and is a set of multiple time series. Many Grafana data sources support this type of data.

Reviewer: ```suggestion
With time series data, the data often contain more than a single series, and is a set of multiple time series. Many Grafana data sources support this type of data.
```

Reviewer: "Data" is plural. I know, it's weird.

## Labels

Each time series in Grafana optionally has labels. labels are set a of key/value pairs for identifying dimensions. Example labels could are `{location=us}` or `{country=us,state=ma,city=boston}`. Within a set of time series, the combination of its name and labels identifies each series. For example, `temperature {country=us,state=ma,city=boston}`.

Different sources of time series data have dimensions stored natively, or common storage patterns that allow the data to be extracted into dimensions.

Time series databases (TSDBs) usually natively support dimensionality. Prometheus also stores dimensions in _labels_. In TSDBs such as Graphite or OpenTSDB the term _tags_ is used instead.

In table databases such SQL, these dimensions are generally the `GROUP BY` parameters of a query.

## Multiple dimensions in table format

In SQL or SQL-like databases that return table responses, additional dimensions usually columns in the query response table.

### Single dimension

For example, consider a query like:

```sql
SELECT BUCKET(StartTime, 1h), AVG(Temperature) AS Temp, Location FROM T
  GROUP BY BUCKET(StartTime, 1h), Location
  ORDER BY time asc
```

> Might return a table with three columns that each respectively have a type of time, number, and string.

Reviewer: ```suggestion
Might return a table with three columns that each respectively have data types time, number, and string.
```
